### PR TITLE
SWATCH-2056 UMB Offering ingestion should write product tag in the Offering table

### PR DESCRIPTION
**Pull request by kartikshahc**
Jira issue: SWATCH-2056

## Description

**Changes:**
- [ ] Refactors Subscription, Offering, Contract and uses a common class to find product tag by sku.
- [ ] New sku_product_tag table with *:1 relation to offering. Since a sku can have multiple product tags we separated out in a separate table.
- [ ] Update swatch-contract with new field and table. Since contract also uses Offering we have to make corresponding changes to the entities and h2 table.

**Business Logic:**
1. metered flag - true
We should match subscriptiondefn for metered products. For those payg products only we will check

- engid -> check if exists. If exists then check for included subscription and remove the included subs. Since engids are many for a sku.
- role -> check if exist. We don't check for included subs since currently roles are unique for each product.
If none of the above exists then check for productname otherwise not.
- productname

2. metered flag - false
- We should match subscriptiondefn for non-metered products. For those non-payg products only we will check

- engid -> check if exists. If exists then check for included subscription and remove the included subs. Since engids are many for a sku.
- role -> check if exist. We don't check for included subs since currently roles are unique for each product
If none of the above exists then check for productname otherwise not.
- productname

## Testing

Done happy path test with IQE steps. More to be done by QE
Happy path scenarios that I thought to test:
1. When sku doesn't exists in Offering table. New skus coming in future
2. When sku does exist in Offering but, offcourse without product_tag. - Current cases
3. In some cases there could be multiple product_tags for a sku so, it should write and retrieve multiple producttags
4. When sku exist and product tag exist but, there is  a new product tag. This should update sku_product_tag table. Have to be creative with this.
5. When there is no change in sku, skip the sku sync
6. Check during contract creation since it gets product_tags by sku
7. Check during event creation by service instance ingress.

### Setup

1.  Run the swatch-core app:
```
RHSM_SUBSCRIPTIONS_ENABLE_SYNCHRONOUS_OPERATIONS=true DEV_MODE=true PROM_URL=http://localhost:8082/api/metrics/v1/telemeter/api/v1
 SWATCH_CONTRACTS_INTERNAL_SERVICE=http://localhost:8882 SERVER_PORT=8000 SPRING_PROFILES_ACTIVE=worker,kafka-queue,api,capacity-ingress,rh-marketplace,kafka-queue,rhsm-conduit ./gradlew clean :bootRun 
```

2. Since stage refresh is going we have make some tweaks and get creative.  Point product_url to qa instead of stage in `application.yaml` 
`url: ${PRODUCT_URL:https://product.qa.api.redhat.com/svcrest/product/v3}`

3. Setting up of IQE environment will be in a separate doc but, setup IQE environment and execute: 
`from smqe_tools import sync_sku_for_capacity`

### Steps and Verification

1.  Case 1: When the sku is not present in offering table. This will create a new entry in sku_product_tag table:
`
sync_sku_for_capacity('MCT3360')
`
**Output:**
INFO:root:PUT request response url: http://localhost:8000/api/rhsm-subscriptions/v1/internal/rpc/offerings/sync/MCT3360, status_code: 200
INFO:root:Response: {
    "detail": "syncResult=FETCHED_AND_SYNCED (Successfully fetched and synced updated value from upstream) for offeringSku=\"MCT3360\"."
}
INFO:root:Offering MCT3360 synced.

`get_product_tags('MCT3360')`
**Output:**
`{'data': ['OpenShift Container Platform']}`

**Note: This doesn't persist or store rhel-for-x86**

2. Case 2: Delete the record from sku_product_tag for `MCT3360` to exercise the current case
**Output:** Will be same as above since it will update the sku with product_tags

3. Case 3: Multiple product_tags for a sku `RH00023`
`sync_sku_for_capacity('RH00023')`
**Output:**
INFO:root:PUT request response url: http://localhost:8000/api/rhsm-subscriptions/v1/internal/rpc/offerings/sync/RH00023, status_code: 200
INFO:root:Response: {
    "detail": "syncResult=FETCHED_AND_SYNCED (Successfully fetched and synced updated value from upstream) for offeringSku=\"RH00023\"."
}
INFO:root:Offering RH00023 synced.

`get_product_tags('RH00023')`
**Output:**
INFO:root:GET request response url: http://localhost:8000/api/rhsm-subscriptions/v1/internal/offerings/RH00023/product_tags, status_code: 200
INFO:root:Response: {
    "data": [
        "rhel-for-x86-ha",
        "RHEL for x86",
        "rhel-for-x86-rs"
    ]
}

`Out[8]: {'data': ['rhel-for-x86-ha', 'RHEL for x86', 'rhel-for-x86-rs']}`

4. Case 4: When sku exist and product tag exist but, there is  a new product tag. For this delete a product_tag from the table that was created from case 3. Delete rhel-for-x86-rs product_tag and assume we are syncing again after adding that as a new yaml file.
` sync_sku_for_capacity('RH00023')`
`get_product_tags('RH00023')`

Output: 
Notice that it creates the new product_tag.

5. Case 5: No change to sku then skip the sku sync
`sync_sku_for_capacity('RH00023')`
**Output:**
INFO:root:PUT request response url: http://localhost:8000/api/rhsm-subscriptions/v1/internal/rpc/offerings/sync/RH00023, status_code: 200
INFO:root:Response: {
    "detail": "syncResult=SKIPPED_MATCHING (Upstream matches stored item, did not sync) for offeringSku=\"RH00023\"."
}
INFO:root:Offering RH00023 synced.

6. Case 6: Contract creation
7. Case 7: Event ingestion

### CODE REVIEW

**Review by kahowell — CHANGES_REQUESTED** [review]
Two aspects to be addressed (discussed directly):
* offerings may (rarely) have more than one product tag
* includedSubscriptions should be consulted

**Comment by ntkathole** [comment]
Hey @kartikshahc 
I am seeing some issue specially related to SKUs such as RH02657HR and MCT3557HR, which should not be considered under rhel-for-x86-els-payg, but are getting considered now. We resolved this issue in the past as apart of https://issues.redhat.com/browse/SWATCH-1840 

Reproducer steps: 

- internal_api_add_sku(sku="RH02657HR", org_id=app.user.identity.org_id, start_day="2024-02-01", end_day="2024-02-29", billing_provider="aws", billing_account_id="9458475893")
- app.rhsm_subscriptions.get_skus_report("rhel-for-x86-els-payg") 
```

[    INFO] [root] REST: http://swatch-api-service.ephemeral-ofhg6k.svc:8000/api/rhsm-subscriptions/v1/subscriptions/products/rhel-for-x86-els-payg

[{'billing_provider': 'aws',
  'capacity': 0,
  'has_infinite_quantity': False,
  'hypervisor_capacity': 0,
  'next_event_date': datetime.datetime(2024, 2, 29, 0, 1, 35, 754000, tzinfo=tzlocal()),
  'next_event_type': 'Subscription End',
  'product_name': 'Red Hat Enterprise Linux Server with Satellite, Premium '
                  '(vCPU, Hourly, On-Demand, Billing)',
  'quantity': 1,
  'service_level': 'Premium',
  'sku': 'RH02657HR',
  'subscriptions': [{'id': '447566', 'number': '8236598'}],
  'total_capacity': 0,
  'usage': ''},
 {'billing_provider': 'aws',
  'capacity': 0,
  'has_infinite_quantity': False,
  'hypervisor_capacity': 0,
  'next_event_date': datetime.datetime(2024, 2, 29, 0, 7, 4, 142000, tzinfo=tzlocal()),
  'next_event_type': 'Subscription End',
  'product_name': 'Red Hat Enterprise Linux, Premium (On-Demand, 1 Virtual '
                  'Node, Hourly, Billing)',
  'quantity': 1,
  'service_level': 'Premium',
  'sku': 'MCT3557HR',
  'subscriptions': [{'id': '710232', 'number': '8925281'}],
  'total_capacity': 0,
  'usage': ''}]
```

**Comment by kartikshahc** [comment]
> Hey @kartikshahc I am seeing some issue specially related to SKUs such as RH02657HR and MCT3557HR, which should not be considered under rhel-for-x86-els-payg, but are getting considered now. We resolved this issue in the past as apart of https://issues.redhat.com/browse/SWATCH-1840
> 
> Reproducer steps:
> 
> * internal_api_add_sku(sku="RH02657HR", org_id=app.user.identity.org_id, start_day="2024-02-01", end_day="2024-02-29", billing_provider="aws", billing_account_id="9458475893")
> * app.rhsm_subscriptions.get_skus_report("rhel-for-x86-els-payg")
> 
> ```
> 
> [    INFO] [root] REST: http://swatch-api-service.ephemeral-ofhg6k.svc:8000/api/rhsm-subscriptions/v1/subscriptions/products/rhel-for-x86-els-payg
> 
> [{'billing_provider': 'aws',
>   'capacity': 0,
>   'has_infinite_quantity': False,
>   'hypervisor_capacity': 0,
>   'next_event_date': datetime.datetime(2024, 2, 29, 0, 1, 35, 754000, tzinfo=tzlocal()),
>   'next_event_type': 'Subscription End',
>   'product_name': 'Red Hat Enterprise Linux Server with Satellite, Premium '
>                   '(vCPU, Hourly, On-Demand, Billing)',
>   'quantity': 1,
>   'service_level': 'Premium',
>   'sku': 'RH02657HR',
>   'subscriptions': [{'id': '447566', 'number': '8236598'}],
>   'total_capacity': 0,
>   'usage': ''},
>  {'billing_provider': 'aws',
>   'capacity': 0,
>   'has_infinite_quantity': False,
>   'hypervisor_capacity': 0,
>   'next_event_date': datetime.datetime(2024, 2, 29, 0, 7, 4, 142000, tzinfo=tzlocal()),
>   'next_event_type': 'Subscription End',
>   'product_name': 'Red Hat Enterprise Linux, Premium (On-Demand, 1 Virtual '
>                   'Node, Hourly, Billing)',
>   'quantity': 1,
>   'service_level': 'Premium',
>   'sku': 'MCT3557HR',
>   'subscriptions': [{'id': '710232', 'number': '8925281'}],
>   'total_capacity': 0,
>   'usage': ''}]
> ```
> 
> ```
> In [28]: get_product_tags("MCT3557HR")
> 2024-02-20 13:46:14.098 [    INFO] [root] GET request response url: http://swatch-subscription-sync-service:8000/api/rhsm-subscriptions/v1/internal/offerings/MCT3557HR/product_tags, status_code: 200
> 2024-02-20 13:46:14.098 [    INFO] [root] Response: {
>     "data": [
>         "rhel-for-x86-eus",
>         "RHEL for x86"
>     ]
> }
> Out[28]: {'data': ['rhel-for-x86-eus', 'RHEL for x86']}
> 
> In [29]: get_product_tags("RH02657HR")
> 2024-02-20 13:46:22.490 [    INFO] [root] GET request response url: http://swatch-subscription-sync-service:8000/api/rhsm-subscriptions/v1/internal/offerings/RH02657HR/product_tags, status_code: 200
> 2024-02-20 13:46:22.490 [    INFO] [root] Response: {
>     "data": [
>         "rhel-for-x86-eus",
>         "RHEL for x86"
>     ]
> }
> Out[29]: {'data': ['rhel-for-x86-eus', 'RHEL for x86']}
> ```

Because we only check by productid https://github.com/RedHatInsights/rhsm-subscriptions/blob/f4b01ab03705a1de87743c250cc7b821a9d836d6/src/main/java/org/candlepin/subscriptions/resource/SubscriptionTableController.java#L258-L278, it appears that the issue that @ntkathole reported is already present in the main branch. The subscription creation and sku capacity doesn't exercise the code changes I have done so, we will make a new card to handle it separately as soon as Nikhil confirms.
We can either reopen https://issues.redhat.com/browse/SWATCH-1840 for product tag `rhel-for-x86-els-payg` instead `eus-payg` since that doesn't exists in product config as mentioned in card description.

**Review by ntkathole — APPROVED** [review]
Looks good 👍 

The issue observed above was also on main. Multiple SKUs tested and those are returning product_tags correctly.

**Comment by kartikshahc** [comment]
> Looks good 👍
> 
> The issue observed above was also on main. Multiple SKUs tested and those are returning product_tags correctly.

Great! Will wait for @kahowell approval on this.

**Review by kahowell — CHANGES_REQUESTED** [review]

File: src/main/java/org/candlepin/subscriptions/product/OfferingSyncController.java
Comment (kartikshahc) [reply src/main/java/org/candlepin/subscriptions/product/OfferingSyncController.java:304]: yes sounds good! I will add else block in which if it returns null then set the producttags to null.

File: src/main/java/org/candlepin/subscriptions/util/ProductOfferingSubscriptionService.java
Comment (kartikshahc) [reply src/main/java/org/candlepin/subscriptions/util/ProductOfferingSubscriptionService.java:38]: Yea I didn't find a better name since this functionality will be used by subscription and offering both. The name you suggested is good so, will change it.

File: src/main/java/org/candlepin/subscriptions/product/OfferingSyncController.java
Comment (kartikshahc) [reply src/main/java/org/candlepin/subscriptions/product/OfferingSyncController.java:304]: Actually with the current code it does remove the product tag. I tested it and doesn't need a code change here.

File: src/main/java/org/candlepin/subscriptions/util/ProductOfferingSubscriptionService.java
Comment (kartikshahc) [reply src/main/java/org/candlepin/subscriptions/util/ProductOfferingSubscriptionService.java]: Done

**Comment by kartikshahc** [comment]
IQE failures are unrelated to my changes! Merging the PR

**Pull request merged**
